master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,470 @@
+func buildWriterPipeline(w io.Writer, mimeType string) (archiveWriter, error) {
+	// DelayWriter delays the first write until we have successfully downloaded
+	// some bytes and are more confident that the download will succeed.
+	dw := newDelayWriter(w, 16*1024)
+	closers := []io.Closer{dw}
+	switch mimeType {
+	case MIMEApplicationGZip:
+		gz := gzip.NewWriter(dw)
+		closers = append(closers, gz)
+
+		tw := tar.NewWriter(gz)
+		closers = append(closers, tw)
+
+		return &tarArchiveWriter{archiveClosers{closers}, tw}, nil
+
+	case MIMEApplicationZip:
+		zw := zip.NewWriter(dw)
+		closers = append(closers, zw)
+
+		return &zipArchiveWriter{archiveClosers{closers}, zw, nil}, nil
+
+	default:
+		return nil, fmt.Errorf(
+			"MIME type must be %s or %s but got %s",
+			MIMEApplicationGZip, MIMEApplicationZip, mimeType)
+	}
+}
+
+func (m *Master) getCheckpointStorageConfig(id uuid.UUID) (
+	*expconf.CheckpointStorageConfig, error) {
+	checkpoint, err := m.db.CheckpointByUUID(id)
+	if err != nil || checkpoint == nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
+	if err != nil {
+		return nil, err
+	}
+	spew.Dump(bytes)
+	spew.Dump(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
+
+	legacyConfig, err := expconf.ParseLegacyConfigJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return ptrs.Ptr(legacyConfig.CheckpointStorage()), nil
+}
+
+func (m *Master) getCheckpoint(c echo.Context, mimeType string) error {
+	args := struct {
+		CheckpointUUID string `path:"checkpoint_uuid"`
+	}{}
+	if err := api.BindArgs(&args, c); err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest,
+			"invalid checkpoint_uuid: "+err.Error())
+	}
+
+	checkpointUUID, err := uuid.Parse(args.CheckpointUUID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest,
+			fmt.Sprintf("unable to parse checkpoint UUID %s: %s",
+				args.CheckpointUUID, err))
+	}
+
+	// Assume a checkpoint always has experiment configs
+	storageConfig, err := m.getCheckpointStorageConfig(checkpointUUID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			fmt.Sprintf("unable to retrieve experiment config for checkpoint %s: %s",
+				args.CheckpointUUID, err.Error()))
+	}
+	if storageConfig == nil {
+		return echo.NewHTTPError(http.StatusNotFound,
+			fmt.Sprintf("checkpoint %s does not exist", args.CheckpointUUID))
+	}
+
+	c.Response().Header().Set(echo.HeaderContentType, mimeType)
+	writerPipe, err := buildWriterPipeline(c.Response(), mimeType)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+	}
+
+	downloader, err := newDownloader(storageConfig, writerPipe, args.CheckpointUUID)
+	err = downloader.download(c.Request().Context())
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			fmt.Sprintf("unable to download checkpoint %s: %s", args.CheckpointUUID, err.Error()))
+	}
+
+	err = writerPipe.Close()
```

Comment:
deferring the close would be ideal, and i think just logging the close error is probably.. ok? wdyt?

Reply:
How does io.Closer work in go?  In C, when you close a writable file descriptor, an error there indicates that the final write didn't actually succeed.

Reply:
Both zip and gzip require Close() to be called to finish writing the whole file. So I think Close() needs to be called before we call c.Response().Flush().

Reply:
ah, so we do need to gather the error from Close, and it's definitely not something we can log and ignore.

We also need to guarantee that Close is called (in general).

I'd guess that `c.Response.Flush()` is unnecessary, because I imagine the response will get flushed by the echo framework eventually.

So I agree with the need to `defer`, only it'll be slightly complex because of the need to detect the error.  Might be time for a named return value.

Reply:
When `Close()` is called, we likely have sent a `200`. Not sure what else we can do other than log it in this case. Of course, if a response has not been sent (committed), we can send the error in the response. 

What's the reason to defer here? It seems easier to handle the error if we don't defer. And if any error happens before we call `Close()`, we probably don't care about flushing (which is all `Close()` does).

Reply:
> When Close() is called, we likely have sent a 200. Not sure what else we can do other than log it in this case

I think it's worth finding out what the Echo framework expects to happen in cases like this.  IMO the echo api clearly expects an API endpoint function to return an error when an error is hit, and so honoring that seems like a bullet-proof solution, even without investigating the edge cases of the echo api.

> What's the reason to defer here?

Idiomatic usage of the interface, I think.  It might not leak resources to not call Close on these particular writers (gzip, tar, and zip writer), but I think it's always best to follow the expectations of the interface you are using.  And these objects have Close methods, which normally must be called in either success or failure cases.

This is an example of what I would call "local understandability".  If I see code which uses a Closer interface, I expect the Close method must be called.  If the close call is deferred, then I don't have to look at any other code to know that resources will be cleaned up.  If the code I'm looking at happens to know exactly which objects back that Closer interface, the code might decide Close doesn't need to be called.  But when reading the code, I can't understand that without reading a bunch more code so that I can verify all the assumptions that make it save to _not_ call Close.  In my opinion, writing code optimized for local understandability pays big dividends in the long run.  It's like the difference between code that is _correct_ vs code that is _obviously correct_.

Reply:
Thanks for explaining the use of defer. That makes sense. I was on the fence whether to adhere to the convention or bend it to reduce complexity. It would be easier if I don't have to call Flush(). That works when I test it, but let me double-check Echo's code.

Reply:
Haven't traced it all the way, but echo.ServeHTTP implements [http.Handler](https://pkg.go.dev/net/http#Handler), which does not require flushing or closing the response (doc does not mention closing or flushing at all). Returning from the handler is sufficient.